### DISCUSSION TITLE
Promise.withResolvers() implementation / mocks

### DISCUSSION THREAD

**Discussion by TheBlindHawk** [vitest-dev/vitest#5512] [Q&A] [answered]
Hello, I've been trying to use [Promise.WithResolvers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers) in my repository, but when I ran the tests I got 
 ```TypeError: Promise.withResolvers is not a function```
I am guessing that since it is a new feature (as of March 2024 according to MDN) it still hasn't been implemented on vitest, so I tried creating mocks for it.
```typescript
// with spyOn (Error: withResolvers does not exist)
vi.spyOn(window.Promise.prototype, 'withResolvers').mockImplementation(() => {
  let resolve, reject
  const promise = new Promise((res, rej) => { resolve = res, reject = rej })
  return { promise, resolve, reject }
})

// with mock (TypeError: Cannot read properties of undefined (reading 'replace'))
vi.mock(window.Promise.withResolvers, () => {
  let resolve, reject
  const promise = new Promise((res, rej) => { resolve = res, reject = rej })
  return { promise, resolve, reject }
})
```

I am guessing spyOn isn't going to work since it only works on functions that already exists.  
As for mock, I've been looking for some examples but I only see examples using a file path string, so it's probably not meant to be used this way.  
But besides the two I couldn't find anything that could help me mock ```Promise.withResolvers```.  

```
vite: "^5.2.8",
vitest: "^1.4.0"
```

Is there any way currently to mock ```Promise.withResolvers```?
Will Promise.WithResolvers mocks be coming in with new versions?
if so when are they approximately planned for?
Any help is appreciated, thanks!

**Comment by hi-ogawa** [comment, marked as answer]
The feature like `Promise.withResolvers` is simply a runtime feature and it's not something "Vitest implements". As seen at the bottom of the same MDN page, it looks like Node simply doesn't have it yet.

I'm not sure whether "mocking" is relevant for your use case, but I think typically people would simply monkey-patch (i.e. polyfill) non-existing global. Have you tried something like this?

```ts
Promise.withResolvers = () => {
  let resolve, reject
  const promise = new Promise((res, rej) => { resolve = res, reject = rej })
  return { promise, resolve, reject }
}
```

If you want to "spy" its function calls, then you could probably do:

```ts
import { vi } from "vitest"
Promise.withResolvers = vi.fn(() => {
  // ... same thing here ...
});
```

Note that `vi.mock` is for "module mocking" so it's not for patching some global object.

**Reply by TheBlindHawk** [reply]
Thank you for the detailed explanation!
The example you provided worked perfectly.
I'll make sure to check further what the issue is beforehand next time, I completely forgot to check node.js compatibility even tho I send the link myself...

**Reply by anton-x-t** [reply]
Thank you very much! Worked very good inside `vitest.setup.ts`.
